### PR TITLE
Fast ISR routine and digital filter

### LINKED ISSUE #24: Write a proper digital filter for directly connected buttons.
https://github.com/ShpakovDmitry/My-old-projects/blob/6e86ebe9eb74b7c9b303172c605751fa2a4893e7/embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c#L159-L239

### PR DESCRIPTION
**Pull request by ShpakovDmitry**
closes #24 and #14 

/cc @krisjans

### CODE REVIEW

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 1]
```diff
@@ -10,7 +12,11 @@
 #define RED_LED PD3
 #define GREEN_LED PD4
 
+static volatile uint32_t msFromReset;
+
 const char morsePhraseToPlay[] = "472";
+const char keysToPress[] = "357";
+#define PASS_LEN sizeof(keysToPress) / sizeof(keysToPress[0])
```
Comment (krisjans) [inline embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: I'm in the mood of renaming this evening.
keysToPress -> PASSWORD? It is constant in your code.
If keysToPress is zero-terminated string, then do you need PASS_LEN? Maybe use MAX_PASS_LEN to set buffer sizes?

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 2]
```diff
@@ -1,7 +1,9 @@
 #include <stdio.h>
+#include <stdbool.h>	// true, false
 #include <avr/io.h>
 #include <util/delay.h>
 #include <avr/interrupt.h>
+#include <stdint.h>		// uint32_t;
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: My, respect! :)

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 3]
```diff
@@ -125,99 +131,143 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
-					i = 1;
-				}
-				else {
-					i = 0;
-					PORTD |= (1 << RED_LED);
-					return;
-				}
-			}
-			
+volatile unsigned char pinB;    // here we store RAW PINB data
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: Yeah, `cachedPinB` is beter.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 4]
```diff
@@ -125,99 +131,143 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+volatile bool flagBtnUpdateValid = false;
+uint32_t debounceTime = 100;
+
+// cyclic buffer implementation
+typedef struct {
+	char buff[PASS_LEN];
+	unsigned char pos;
+	unsigned char size;
+}CyclicBuffer;
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: Nice, thank you, did not even noticed this moment. I thought, that order does not matter as we should know such buff size at compile time or will get compilation-error.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 5]
```diff
@@ -125,99 +131,143 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+
+CyclicBuffer cyclicBuffer;
+
+void initCyclicBuffer(CyclicBuffer *buff) {
+	buff->pos = 0;
+	buff->size = PASS_LEN;
+}
+
+void addCharToCyclicBuffer(CyclicBuffer *buff, const char ch) {
+	*(buff->buff + buff->pos) = ch;
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: My old coding style returned due to copy-paste. For me when i wrote this `*(buff->buff + buff->pos) = ch; ` looked better to understand on pointer level.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 6]
```diff
@@ -125,99 +131,143 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+			if ( (msFromReset - buttons[i].lastDebounceTime ) > debounceTime) {
+				if (buttonTmp != buttons[i].currState) {
+					buttons[i].currState = buttonTmp;
+					if (buttons[i].currState == PRESSED) {
+						addCharToCyclicBuffer(&cyclicBuffer, buttons[i].keyChar);
+					}
+				}
+			}
+			buttons[i].lastState = buttonTmp;
+        }
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: Yes, releasing button is crucial, so one pressed button at a time. Will introduce correct button handling in next commits. So method how to handle keys press will become less over engineered. 
This bit trick is nice, also did not know it. Will use it.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 7]
```diff
@@ -125,99 +131,143 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+	// enable output compare A match interrupt
+	TIMSK |= (1 << OCIE0A);
+	// start counting ms at this point; actually when glob interrupt flag is set
+	msFromReset = 0;   
+}
+
+void initContoroller(void) {
+	initIO();
+	initTimer0();
 	SREG |= (1 << 7);
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: Yes, actually there is `sei()` marco which comes with `avr-libc`. Will change in next commits.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 8]
```diff
@@ -125,99 +131,143 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+	initCyclicBuffer(&cyclicBuffer);
 	
 	while(1) {
-		;
+		updateButtonState();
+		if (findSequenceInCyclicBuffer(&cyclicBuffer, keysToPress) == true) {
+			for (uint8_t i = 0; i < MAX_REPEAT; i++) {
+				playMorsePhrase(morsePhraseToPlay);
+			}
+		}
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: No, it will play endless, until any key will be pressed. Will correct in next commits.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 9]
```diff
@@ -125,99 +131,143 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+    if (flagBtnUpdateValid == true) {
+		// low-pass-filter using debounce time
+        for (unsigned char i = 0; i < 8; i++) {
+			uint8_t buttonTmp = ( pinB & (1 << buttons[i].pin) ) ? RELEASED : PRESSED;
+			
+			if (buttonTmp != buttons[i].lastState) {
+				buttons[i].lastDebounceTime = msFromReset;
+			}
+			
+			if ( (msFromReset - buttons[i].lastDebounceTime ) > debounceTime) {
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: Will add handler for this event.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c (hunk 1 above)
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: Good will use this.
Actually i like Your comments a lot. They are very significant for me. I am actually learning and improving my skill from them. Thank you.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 10]
```diff
@@ -125,99 +131,143 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+
+void addCharToCyclicBuffer(CyclicBuffer *buff, const char ch) {
+	*(buff->buff + buff->pos) = ch;
+	if (buff->pos == buff->size - 1) {
+		buff->pos = 0;
 	}
 	else {
-		i = 0;
+		buff->pos += 1;
 	}
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: Agree, this is more simpler and nice looking way.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (krisjans) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: If you know buffer size at compile time, then you don't need "size" element at all.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 11]
```diff
@@ -10,7 +13,11 @@
 #define RED_LED PD3
 #define GREEN_LED PD4
 
+static volatile uint32_t msFromReset;
+
 const char morsePhraseToPlay[] = "472";
+const char PASSWORD[] = "357";
+#define MAX_PASS_LEN 5
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: This is good, as compiler will tell us if when things goes wrong.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 12]
```diff
@@ -125,99 +132,182 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+typedef struct {
+	unsigned char pos;
+	unsigned char size;
+	char buff[MAX_PASS_LEN];
+}CyclicBuffer;
+
+// functions to handle cyclic buffer
+void initCyclicBuffer(CyclicBuffer *buff);
+void addCharToCyclicBuffer(CyclicBuffer *buff, const char ch);
+bool findSequenceInCyclicBuffer(CyclicBuffer *buff, const char *seq);
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: Actually this lines were copy-pasted, but it was meant that in future commits there will be separate *.c and *.h files.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 13]
```diff
@@ -125,99 +132,182 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+
+CyclicBuffer cyclicBuffer;
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: This makes more sense what it holds.

File: embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c [hunk 14]
```diff
@@ -125,99 +132,182 @@ void playMorsePhrase (const char* morsePhraseToPlay) {
...
+
+void initCyclicBuffer(CyclicBuffer *buff) {
+	buff->pos = 0;
+	buff->size = MAX_PASS_LEN;
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: Agree, about resource waste. At that time i thought that it will make code more readable(at least for me), as buffer size is inside buffer struct, and not defined anywhere else.
So, not to waste resources it is possible to determine size from `MAX_PASS_LEN` macro and so `CyclicBuffer` struct may become:
```c
typedef struct {
	unsigned char pos;
	char buff[MAX_PASS_LEN];
}CyclicBuffer;
```
And everywhere in code use MAX_PASS_LEN as common buffer size.

What are the benefits of the code you introduced? 
Does it makes `CyclicBuffer` init routine more universal?

[the remark this answers is not in the document]
Comment (krisjans) [reply embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c]: I'm so glad to see counter-questions from you!
TL;DR - I was wrong. Long story below:

My intended reason behind that code: de-couple buffer size from structure definition so that I could write unit-tests for different buffer sizes. And to actually do that I must omit buffer size in structure definition:
```C
typedef struct {
	unsigned char size;
	unsigned char pos;
	char buff[];
} CyclicBuffer;
```
This works nice for run-time allocation:
```C
void doSomeStuff(unsigned char buffSize) {
     CyclicBuffer *cb = initCyclicBuffer(malloc(sizeof(CyclicBuffer) + buffSize), buffSize);
}
```
**BUT** to allocate buffer **with specific size** at **compile-time** becomes a pain:
```C
char cb2_data[MAX_PASS_LEN + sizeof(CyclicBuffer)];
CyclicBuffer * cb2 = (CyclicBuffer *)cb2_data;
void main() {
    initCyclicBuffer(cb2, sizeof(cb2_data));
```
And now it became really inefficient - I'm forced to use pointer.
How about using union to pack CyclicBuffer and data buffer in one space?
```C
union {
	CyclicBuffer h;
	char data[MAX_PASS_LEN + sizeof(CyclicBuffer)];
} cb2;
void main() {
    initCyclicBuffer(&cb2.h, sizeof(cb2));
```
This is kind of better, but still - I force myself to use magic "h" (stand for "header") union member. And one line definiton is now bloated to to 4 lines 
Some other reasons I was going to use to justify my megacode code:
- allows many CyclicBuffer-s with different size each.
- create buffers at run-time.
- re-use any chunk of memory as CyclicBuffer.

Is this needed for this particular project - no. 
Maybe my code could be justified in larger project, where many buffers with different sizes needs to be allocated dynamically, but when one global buffer is needed - it is just a bloat.

p.s. In programming such buffers often are called "Circular" or "Round" buffers. Write position is called "head" and read position is called "tail" (worm/snake analogy).

File: embedded/EscapeROOM/radio_morse_panel/source/misc.c [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+		pos++;
+	}
+	return pos;
+}
+
+uint32_t getDeltaTime (uint32_t t1, uint32_t t0) {
+	return t1 - t0;
+}
+
+void updateButtonState(void) {
```
Comment (krisjans) [inline embedded/EscapeROOM/radio_morse_panel/source/misc.c:58]: Following comment is my personal preference, not a bug or change request.
In misc.c I thend to put small helper functions, that does not belong to any specific object and are re-used from several other .c files. Like isOneBitSet() or getSetBitPosition().
updateButtonState() is not a helper in my opinion.  It belongs to object (.c file ) that deals with buttons and/or user input. I would put all button related functions into separate file - button.c.

**Comment by ShpakovDmitry** [comment]
Yeah, i was inspired by that one stm8 project :). For now, in my opinion, organizing project, like all ISR goes to interrupt.c, init routines goes to init.c and etc, may help to better ''organize'' small project when there is no lib's used in comparison with project where all routines are in one main.c file maiking code mess. I guess that there is no any standard how to organize project, all relies on personal taste or project main guidelines. I personally agree, that it would be time consuming when reusing and adopting code for another projects, if organized like this project. So in future i tend to use encapsulated lib's for every peripheral, structure, algorithm and etc, giving only API to main code and hiding all internal structure and direct access to data. Also i believe that hardware specific routines may be implemented separately for, example in hardware.c, to lead to code portability to another platforms. But, maybe, this is true for larger projects. I think it is not worth for small of them, because it may become development time and firmware size consuming. I know that linker is smart enough to throw away unused routines to keep obj file smaller, but the question may become - why we need, for example, include `heavy` GPIO lib to set pin high/low if it could be done in one line of code. I think it is not worth of it. I admit that i may be wrong at any point of this, i am only learning design patterns at the moment.

File: embedded/EscapeROOM/radio_morse_panel/source/misc.c (hunk 15 above)
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/source/misc.c:58]: Taking this into account in future. Thanks.

File: embedded/EscapeROOM/radio_morse_panel/include/globals.h [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+#include <string.h>
+#include "cyclic_buffer.h"
+
+#define DEBOUNCE_TIME 100
+#define DOT_DURATION 150
+#define MAX_REPEAT 5
+#define RED_LED PD3
+#define GREEN_LED PD4
+
+static const char morsePhraseToPlay[] = "472";
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/include/globals.h:19]: I missed this moment. Fix in next commits.

File: embedded/EscapeROOM/radio_morse_panel/include/globals.h [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+static const char PASSWORD[MAX_BUFF_LEN] = "357\0";
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/include/globals.h:20]: Using `extern` for variables in `globals.h`

File: embedded/EscapeROOM/radio_morse_panel/README.md [hunk 18]
```diff
@@ -1,4 +1,13 @@
 ### Radio_morse_panel project prototype
-Input puttons and output leds with speaker are marked with
+To build project execute:
+```bash
+make
+```
+To check controller memory usage by firmware execute:
+```bash
+make
+avr-size ./build/radio_morse.o --format=avr --mcu=attiny2313
```
[the remark this answers is not in the document]
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/README.md:9]: Right, did not think of that.

File: embedded/EscapeROOM/radio_morse_panel/include/button.h [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+static Button buttons[] = {
+	{ PB0, '1'},
+	{ PB1, '2'},
+	{ PB2, '3'},
+	{ PB3, '4'},
+	{ PB4, '5'},
+	{ PB5, '6'},
+	{ PB6, '7'},
+	{ PB7, '8'}
+};
```
Comment (krisjans) [inline embedded/EscapeROOM/radio_morse_panel/include/button.h:34]: const?
And static variables should not be in header files.

File: embedded/EscapeROOM/radio_morse_panel/source/button.c [hunk 20]
```diff
@@ -0,0 +1,38 @@
+#include <stdint.h>
+#include "globals.h"
+#include "button.h"
+#include "misc.h"
+
+CyclicBuffer pressedButtons;
+
+void updateButtonState(void) {
```
Comment (krisjans) [inline embedded/EscapeROOM/radio_morse_panel/source/button.c:8]: You moved the function definition, but forgot to move declaration from misc.h to button.h

**Comment by krisjans** [comment]
> [....] Also i believe that hardware specific routines may be implemented separately for, example in hardware.c, to lead to code portability to another platforms. But, maybe, this is true for larger projects. I think it is not worth for small of them, because it may become development time and firmware size consuming. I know that linker is smart enough to throw away unused routines to keep obj file smaller, **but the question may become - why we need, for example, include `heavy` GPIO lib to set pin high/low if it could be done in one line of code.** I think it is not worth of it. I admit that i may be wrong at any point of this, i am only learning design patterns at the moment.

Valid question. In my projects I try to avoid heavy libs and use C macros instead. Something like this:

gpio.h:
```C
#define SET_GPIO_DIR_OUT(dir, bit) (dir) |= (1 << (bit))
#define SET_GPIO_DIR_IN(dir, bit) (dir) &= ~(1 << (bit))
#define GET_GPIO_INPUT(port, bit) (port) & (1 << (bit))
#define SET_GPIO_HI(port, bit) (port) |= (1 << (bit))
#define SET_GPIO_LO(port, bit) (port) &= ~(1 << (bit))
```
my_project_hw_config.h
```C
#include "gpio.h"
#define BUTTON_UP_DIR          DIRA
#define BUTTON_UP_PORT         PORTA
#define BUTTON_UP_BIT          3
#define BUTTON_DOWN_DIR        DIRC
#define BUTTON_DOWN_PORT       PORTC
#define BUTTON_DOWN_BIT        7
#define INIT_BUTTON_UP() SET_GPIO_DIR_IN(BUTTON_UP_DIR, BUTTON_UP_BIT)
#define INIT_BUTTON_DOWN() SET_GPIO_DIR_IN(BUTTON_DOWN_DIR, BUTTON_UP_BIT)
#define BUTTON_UP_PRESSED() GET_GPIO_INPUT(BUTTON_UP_PORT, BUTTON_UP_BIT)
#define BUTTON_DOWN_PRESSED() GET_GPIO_INPUT(BUTTON_DOWN_PORT, BUTTON_DOWN_BIT)
```
main.c
```C
#include "my_project_hw_config.h"

void main(void) {
    INIT_BUTTON_UP();
    INIT_BUTTON_DOWN();
    if (BUTTON_UP_PRESSED()) {
        printf("Up button pressed\n");
    }
}
```
I easily move any button to any gpio. I can even easily migrate to different micro-controller. And at the same time I have library-like interface.

File: embedded/EscapeROOM/radio_morse_panel/include/button.h (hunk 19 above)
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/include/button.h]: Clear.
Regarding to the header files. This is great, will try to use in future this approach. Thank you!

File: embedded/EscapeROOM/radio_morse_panel/source/button.c (hunk 20 above)
Comment (ShpakovDmitry) [reply embedded/EscapeROOM/radio_morse_panel/source/button.c:22]: Correcting in next commits.

**Comment by ShpakovDmitry** [comment]
https://github.com/ShpakovDmitry/My-old-projects/blob/af4d1cfd17fc5a898bc3e63f003f46971c169761/embedded/EscapeROOM/radio_morse_panel/source/morse.c#L10-L20
To fully use the gpio related defines `morse.c` need to be changed. Will change in next commits.
